So, even the old version 6.0.1 does NOT work with JDK16, so this package could actually be considered broken right now. The only reason it installed properly for me was that my nixpkgs channel wasn't updated yet.

I tested this on a different machine and could confirm this:

<details>
  <summary>Click to expand!</summary>

```bash
# nixpkgs wasn't updated yet
$ NIXPKGS_ALLOW_UNFREE=1 nix-env -iA nixpkgs.stm32cubemx
[...]
mkdir: created directory '/nix/store/b09nvyxa68n3f8zsqbn2l9s9gqgmksq6-stm32cubemx-6.0.1/share/icons'
[...]
$ nix-store -q  --references /nix/store/b09nvyxa68n3f8zsqbn2l9s9gqgmksq6-stm32cubemx-6.0.1
/nix/store/cwnwyy82wrq53820z6yg7869z8dl5s7g-bash-4.4-p23
/nix/store/v8dykc8xwq8f0gv3pcca4df8j8kvqsga-openjdk-15.0.1-ga
/nix/store/xir9arx16wlig1v0wikp31gwwjy6l0n2-stm32CubeMX.desktop
/nix/store/b09nvyxa68n3f8zsqbn2l9s9gqgmksq6-stm32cubemx-6.0.1
$ /nix/store/b09nvyxa68n3f8zsqbn2l9s9gqgmksq6-stm32cubemx-6.0.1/bin/stm32cubemx
[...] # lots of output, works perfectly
$ nix-channel --list
nixpkgs https://nixos.org/channels/nixpkgs-unstable
$ nix-channel --update
unpacking channels...
warning: Nix search path entry '/nix/var/nix/profiles/per-user/root/channels' does not exist, ignoring
created 1 symlinks in user environment
$ NIXPKGS_ALLOW_UNFREE=1 nix-env -iA nixpkgs.stm32cubemx
[...]
mkdir: created directory '/nix/store/n19vahbsrfhgmbj4kx1ws98npmjw6wsv-stm32cubemx-6.0.1/share/icons'
[...]
created 45 symlinks in user environment
$ nix-store -q  --references /nix/store/b09nvyxa68n3f8zsqbn2l9s9gqgmksq6-stm32cubemx-6.0.1
/nix/store/b8pk4qfj00rppf3nn0jzqi5va5d543mp-openjdk-16+36
/nix/store/bhqm8m0svacc4ps1hbgqhqa3iz4g3256-stm32CubeMX.desktop
/nix/store/jdi2v7ir1sr6vp7pc5x0nhb6lpcmg6xg-bash-4.4-p23
/nix/store/n19vahbsrfhgmbj4kx1ws98npmjw6wsv-stm32cubemx-6.0.1
$ /nix/store/n19vahbsrfhgmbj4kx1ws98npmjw6wsv-stm32cubemx-6.0.1/bin/stm32cubemx
[...] # lots of output, and the splashscreen appears, but:
Exception in thread "AWT-EventQueue-0" java.lang.ExceptionInInitializerError
	at com.thoughtworks.xstream.XStream.setupConverters(XStream.java:990)
	at com.thoughtworks.xstream.XStream.<init>(XStream.java:593)
	at com.thoughtworks.xstream.XStream.<init>(XStream.java:515)
	at com.thoughtworks.xstream.XStream.<init>(XStream.java:484)
	at com.thoughtworks.xstream.XStream.<init>(XStream.java:430)
	at com.thoughtworks.xstream.XStream.<init>(XStream.java:397)
	at com.st.library.rawvaluescontainer.FilesUtils.<clinit>(FilesUtils.java:41)
	at com.st.library.analyticscore.AnalyticsContext.<init>(AnalyticsContext.java:167)
	at com.st.library.dynamicproxy.AnalyticsInvocationHandler.<init>(AnalyticsInvocationHandler.java:36)
	at com.st.library.builder.AnalyticsBuilder.getAnalytics(AnalyticsBuilder.java:33)
	at com.st.library.analyticscore.AnalyticsAPI.initialize(AnalyticsAPI.java:132)
	at com.st.microxplorer.plugins.analytics.AnalyticsPlugin.<init>(AnalyticsPlugin.java:165)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:78)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:499)
	at java.base/java.lang.reflect.ReflectAccess.newInstance(ReflectAccess.java:128)
	at java.base/jdk.internal.reflect.ReflectionFactory.newInstance(ReflectionFactory.java:350)
	at java.base/java.lang.Class.newInstance(Class.java:642)
	at com.st.microxplorer.plugin.PluginManage.loadPluginJar(PluginManage.java:335)
	at com.st.microxplorer.plugin.PluginManage.PluginSearch(PluginManage.java:206)
	at com.st.microxplorer.plugin.PluginManage.PluginDiscovery(PluginManage.java:49)
	at com.st.microxplorer.maingui.MainPanel.initialization(MainPanel.java:236)
	at com.st.microxplorer.maingui.MainFrame.initialization(MainFrame.java:149)
	at com.st.microxplorer.maingui.MicroXplorer.run(MicroXplorer.java:226)
	at com.st.microxplorer.maingui.STM32CubeMX.run(STM32CubeMX.java:25)
	at java.desktop/java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:316)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:770)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:721)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:715)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:391)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:740)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
Caused by: java.lang.reflect.InaccessibleObjectException: Unable to make field private final java.util.Comparator java.util.TreeMap.comparator accessible: module java.base does not "opens java.util" to unnamed module @13663f96
	at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:357)
	at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:297)
	at java.base/java.lang.reflect.Field.checkCanSetAccessible(Field.java:177)
	at java.base/java.lang.reflect.Field.setAccessible(Field.java:171)
	at com.thoughtworks.xstream.core.util.Fields.locate(Fields.java:40)
	at com.thoughtworks.xstream.converters.collections.TreeMapConverter.<clinit>(TreeMapConverter.java:50)
	... 39 more
```

</details>

No idea what the actual error means, but it's probably not fixable on our end.

I tried it with jdk11 as suggested, and that seems to work perfectly 🎉

I also sent [a bug report to ST](https://community.st.com/s/question/0D53W00000jnOzPSAU/stm32cubemx-crashes-on-launch-with-openjdk16).

> Result of `nixpkgs-review pr 120341` at [91c7545](https://github.com/NixOS/nixpkgs/commit/91c75455a89a82c887e9804b7d28637d8a11720e) run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
> 1 package built successfully:
> 1 suggestion:
> 
>     * warning: unused-argument
>       Unused argument: `requireFile`.
>       Near pkgs/development/tools/misc/stm32cubemx/default.nix:1:16:
>       ```
>         |
>       1 | { lib, stdenv, requireFile, makeDesktopItem, icoutils, fdupes, imagemagick, jre, fetchzip }:
>         |                ^
>       ```

I applied this change as well, works perfectly still.

--- a/pkgs/development/tools/misc/stm32cubemx/default.nix
+++ b/pkgs/development/tools/misc/stm32cubemx/default.nix
@@ -1,52 +1,57 @@
-{ lib, stdenv, requireFile, makeDesktopItem, libicns, imagemagick, jre, fetchzip }:
-
+{ lib, stdenv, makeDesktopItem, copyDesktopItems, icoutils, fdupes, imagemagick, jdk11, fetchzip }:
+# TODO: JDK16 causes STM32CubeMX to crash right now, so we fixed the version to JDK11
+# This may be fixed in a future version of STM32CubeMX. This issue has been reported to ST:
+# https://community.st.com/s/question/0D53W00000jnOzPSAU/stm32cubemx-crashes-on-launch-with-openjdk16
+# If you're updating this derivation, check the link above to see if it's been fixed upstream
+# and try replacing all occurrences of jdk11 with jre and test whether it works.
 let
-  version = "6.0.1";
-  desktopItem = makeDesktopItem {
-    name = "stm32CubeMX";
-    exec = "stm32cubemx";
-    desktopName = "STM32CubeMX";
-    categories = "Development;";
-    icon = "stm32cubemx";
-  };
+  iconame = "STM32CubeMX";
 in
 stdenv.mkDerivation rec {
   pname = "stm32cubemx";
-  inherit version;
-
+  version = "6.2.1";
 
   src = fetchzip {
-    url = "https://sw-center.st.com/packs/resource/library/stm32cube_mx_v${builtins.replaceStrings ["."] [""] version}.zip";
-    sha256 = "15vxca1pgpgxgiz4wisrw0lylffdwnn4n46z9n0q37f8hmzlrk8f";
-    stripRoot= false;
+    url = "https://sw-center.st.com/packs/resource/library/stm32cube_mx_v${builtins.replaceStrings ["."] [""] version}-lin.zip";
+    sha256 = "0m5h01iq0mgrr9svj4gmykfi9lsyjpqzrkvlizff26c8dqad59c5";
+    stripRoot = false;
   };
 
-  nativeBuildInputs = [ libicns imagemagick ];
+  nativeBuildInputs = [ icoutils fdupes imagemagick copyDesktopItems];
+  desktopItems = [
+    (makeDesktopItem {
+      name = "stm32CubeMX";
+      exec = "stm32cubemx";
+      desktopName = "STM32CubeMX";
+      categories = "Development;";
+      comment = "STM32Cube initialization code generator";
+      icon = "stm32cubemx";
+    })
+  ];
 
   buildCommand = ''
-    mkdir -p $out/{bin,opt/STM32CubeMX,share/applications}
-    cp -r $src/. $out/opt/STM32CubeMX/
-    chmod +rx $out/opt/STM32CubeMX/STM32CubeMX.exe
+    mkdir -p $out/{bin,opt/STM32CubeMX}
+    cp -r $src/MX/. $out/opt/STM32CubeMX/
+    chmod +rx $out/opt/STM32CubeMX/STM32CubeMX
     cat << EOF > $out/bin/${pname}
     #!${stdenv.shell}
-    ${jre}/bin/java -jar $out/opt/STM32CubeMX/STM32CubeMX.exe
+    ${jdk11}/bin/java -jar $out/opt/STM32CubeMX/STM32CubeMX
     EOF
     chmod +x $out/bin/${pname}
 
-    icns2png --extract $out/opt/STM32CubeMX/${pname}.icns
+    icotool --extract $out/opt/STM32CubeMX/help/${iconame}.ico
+    fdupes -dN . > /dev/null
     ls
     for size in 16 24 32 48 64 128 256; do
       mkdir -pv $out/share/icons/hicolor/"$size"x"$size"/apps
-      if [ -e ${pname}_"$size"x"$size"x32.png ]; then
-        mv ${pname}_"$size"x"$size"x32.png \
+      if [ $size -eq 256 ]; then
+        mv ${iconame}_*_"$size"x"$size"x32.png \
           $out/share/icons/hicolor/"$size"x"$size"/apps/${pname}.png
       else
-        convert -resize "$size"x"$size" ${pname}_256x256x32.png \
+        convert -resize "$size"x"$size" ${iconame}_*_256x256x32.png \
           $out/share/icons/hicolor/"$size"x"$size"/apps/${pname}.png
       fi
     done;
-
-    ln -s ${desktopItem}/share/applications/* $out/share/applications
   '';
 
   meta = with lib; {

